kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala

```diff
@@ -757,87 +760,327 @@ object Resource extends ResourceInstances with ResourcePlatform {
+   *   b) if the fiber is succeeded, run its finalizers
+   * 2. If the fiber is canceled or errored, finalize
+   * 3. If the fiber succeeds, await joining and extend finalizers into joining scope
+   * 4. If a fiber is canceled *after* it completes, we explicitly snag the finalizers and run them, setting the result to Canceled()
+   *   a) Note that we are missing a check here! We need to disable this behavior if someone has already joined
+   */
+  def start[A](fa: Resource[F, A]): Resource[F, Fiber[Resource[F, *], Throwable, A]] = {
+    final case class State(
+        fin: F[Unit] = F.unit,
+        runOnComplete: Boolean = false,
+        completed: Boolean = false,
+        canceled: Boolean = false)
+
+    Resource {
+      import Outcome._
+
+      F.ref[State](State()) flatMap { state =>
+        val folded =
+          fa.fold(_.pure[F], release => state.update(s => s.copy(fin = s.fin !> release)))
+
+        val finalized = F.guaranteeCase(folded) {
+          case Canceled() | Errored(_) =>
+            state.get.flatMap(_.fin)
+
+          case Succeeded(_) =>
+            val maybeFins = state modify { s =>
+              (
+                s.copy(completed = true),
+                Some(s.fin).filter(_ => s.runOnComplete && !s.canceled))
+            }
+
+            maybeFins.flatMap(_.getOrElse(F.unit))
+        }
+
+        F.start(finalized) map { outer =>
+          val fiber = new Fiber[Resource[F, *], Throwable, A] {
+
+            def cancel =
+              Resource eval {
+                F uncancelable { poll =>
+                  // technically cancel is uncancelable, but separation of concerns and what not
+                  val maybeFins = poll(outer.cancel) *> {
+                    state modify { s =>
+                      (s.copy(canceled = true), Some(s.fin).filter(_ => s.completed))
+                    }
+                  }
+
+                  maybeFins.flatMap(_.getOrElse(F.unit))
+                }
+              }
+
+            def join =
+              Resource eval {
+                val results = outer.join.map[Outcome[Resource[F, *], Throwable, A]] {
+                  case Canceled() => Canceled()
+                  case Errored(e) => Errored(e)
+
+                  case Succeeded(fp) =>
+                    Succeeded(Resource.make(fp)(_ => state.get.flatMap(_.fin)))
+                }
+
+                // we do this after rather than before to avoid breaking backpressure
+                results flatMap { oc =>
+                  F.ifF(state.get.map(_.canceled))(
+                    Outcome.Canceled[Resource[F, *], Throwable, A](),
+                    oc)
+                }
+              }
+          }
+
+          val finalizeEarly = {
+            val maybeFins = state modify { s =>
+              (
+                s.copy(runOnComplete = true),
+                Some(s.fin).filter(_ => s.completed && !s.canceled))
```

Comment:
With the change to `State`, this could become
```suggestion
                s.fin.traverse_(x => x)
```